src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+-- iteration-space control path that got there.
+--
+-- Another way to see this is that the compilation of At is the only
+-- thing that cares about and alters the type of the reference being
+-- passed around.  Ergo, all the At constructors must occur in the
+-- same order.
+
+-- === Use smart constructors ===
+
+-- We maintain a normal form for Use data with these smart
+-- constructors.  The main invariant is
+--
+-- - Max constructors appear outside of adjacent Sum constructors.
+--
+-- We also take the opportunity to peephole-simplify the Use
+-- structure.
+
+instance MaxPlus (Use iter) where
+  zero = Const zero
+
+  max (Const Unbounded) _ = Const Unbounded
+  max _ (Const Unbounded) = Const Unbounded
+  max (Const b1) (Const b2) = Const $ b1 `max` b2
+  max (Max uses1) (Max uses2) = Max $ uses1 ++ uses2
+  -- TODO This is trying to eliminate zero-valued elements, because
+  -- they can cause (potentially severe) loss of precision downstream.
+  -- However, that relies on upstream construction canonicalizing all
+  -- semantically-zero Uses to the `zero` use, which they may not
+  -- entirely do.
+  max (Max uses1) use2
+    | use2 == zero = Max uses1
+    | otherwise = Max $ uses1 ++ [use2]
+  max use1 (Max uses2)
+    | use1 == zero = Max uses2
+    | otherwise = Max $ use1:uses2
+  max use1 use2
+    | use1 == zero = use2
+    | use2 == zero = use1
+    | otherwise = Max [use1, use2]
+
+  plus (Const Unbounded) _ = Const Unbounded
+  plus _ (Const Unbounded) = Const Unbounded
+  plus (Const b1) (Const b2) = Const $ b1 `plus` b2
+  plus (Max uses1) (Max uses2) = Max uses' where
+    uses' = concat $ flip map uses1 (\use1 ->
+      flip map uses2 \use2 -> plus use1 use2)
+      -- TODO: Detect overlapping uses here?
+  plus u1@(Max _) use2
+    | use2 == zero = u1
+    | otherwise = plus u1 (Max [use2])
+  plus use1 u2@(Max _)
+    | use1 == zero = u2
+    | otherwise = plus (Max [use1]) u2
+  plus (Sum uses1) (Sum uses2) = Sum $ uses1 ++ uses2
+  plus (Sum uses1) use2
+    -- TODO Is there a way to deduplicate these zero checks?
+    | use2 == zero = Sum uses1
+    | otherwise = Sum $ uses1 ++ [use2]
+  plus use1 (Sum uses2)
+    | use1 == zero = Sum uses2
+    | otherwise = Sum $ use1:uses2
+  plus use1 use2
+    | use1 == zero = use2
+    | use2 == zero = use1
+    | otherwise = Sum [use1, use2]
+
+-- === Interpreting Access into Use ===
+
+-- The main interpreter is pretty straightforward.  The main trick is
+-- just semantics: an indexing expression becomes a (nested) `At`,
+-- which represents a data-space loop that checks that the data-space
+-- variable equals the indexing expression.
+interp :: Access iter -> Use iter
+interp (Location [] ct) = Const ct
+interp (Location (expr:exprs) ct) = At expr rest where
+  rest = interp $ Location exprs ct
+interp (Any accesses) = foldl max zero uses where
+  uses = map interp accesses
+interp (All accesses) = foldl plus zero uses where
+  uses = map interp accesses
```

Comment:
Perhaps just `$ map interp accesses` and similarly in all cases above, to avoid the extra hang? It might also be more readable if you use lambda case (since you won't need to mention `interp` so many times):
```
interp = \case
  Location [] ct -> Const ct
  Location (expr: exprs) ct -> ...
  ...
```

Reply:
Thanks!